File: src/main/scala/com/github/jarol/azure/search/spark/sql/connector/read/SchemaCompatibilityException.scala
package com.github.jarol.azure.search.spark.sql.connector.read

import com.azure.search.documents.indexes.models.SearchField
import com.github.jarol.azure.search.spark.sql.connector.AzureSparkException
import org.apache.spark.sql.types.StructField

class SchemaCompatibilityException(missingFields: Seq[String],
                                   nonCompatibleFields: Map[StructField, SearchField])
  extends AzureSparkException {

  override def getMessage: String = {

    val missingFieldDescription: Option[String] = if (missingFields.isEmpty) None else {
      Some(
        s"${missingFields.size} missing fields " +
          s"${missingFields.mkString("(", ",", ")")})"
      )
    }

    val nonCompatibleFieldsDescription: Option[String] = if (nonCompatibleFields.isEmpty) None else {

      val tuplesDescription = nonCompatibleFields.map {
        case (k, v) => s"${k.name} (spark type: ${k.dataType.typeName}, search type: ${v.getType})"
      }.mkString("(", ",", ")")

      Some(
        s"${nonCompatibleFields.size} fields with non-compatible data types $tuplesDescription"
      )
    }

    (missingFieldDescription, nonCompatibleFieldsDescription) match {
      case (Some(a), Some(b)) => s"Found $a and $b"
      case (Some(a), None) => s"Found $a"
      case (None, Some(b)) => s"Found $b"
      case _ => "No missing or non-compatible fields"
    }
  }
}

File: core/src/main/java/com/github/jarol/azure/search/spark/sql/connector/core/schema/conversion/SchemaCompatibilityException.java
package com.github.jarol.azure.search.spark.sql.connector.core.schema.conversion;

import com.azure.search.documents.indexes.models.SearchField;
import org.apache.spark.sql.types.StructField;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Exception raised when a Schema incompatibility is met
 * <p>
 * It may happen due to
 * <ul>
 *     <li>a non existing schema field (i.e. a schema field that does not exist on a Search index)</li>
 *     <li>an existing schema field with incompatible datatype with respect to its namesake Search field</li>
 * </ul>
 */

public class SchemaCompatibilityException
        extends IllegalStateException {

    /**
     * Create an instance
     * @param message exception message
     */

    public SchemaCompatibilityException(
            @NotNull String message
    ) {
        super(message);
    }

    /**
     * Create an instance due to missing schema fields
     * @param index index name
     * @param missingFields name of non-existing fields
     * @return an exception instance
     */

    public static @NotNull SchemaCompatibilityException forMissingFields(
            String index,
            @NotNull List<String> missingFields
    ) {

        String message = String.format("%s schema fields (%s) do not exist on index %s",
                missingFields.size(),
                String.join(", ", missingFields),
                index
        );
        return new SchemaCompatibilityException(message);
    }

    @Contract("_ -> new")
    public static @NotNull SchemaCompatibilityException forNonCompatibleFields(
            @NotNull Map<StructField, SearchField> fieldMap
    ) {

        String description = fieldMap.entrySet().stream().map(
                entry -> String.format(
                        "%s (%s, not compatible with %s)",
                        entry.getKey().name(),
                        entry.getKey().dataType(),
                        entry.getValue().getType()
                )
        ).collect(Collectors.joining(", "));

        String message = String.format(
                "%s fields have incompatible data types (%s)",
                fieldMap.size(),
                description
                );

        return new SchemaCompatibilityException(message);
    }

    @Contract("_ -> new")
    public static @NotNull SchemaCompatibilityException forSchemaFieldsWithoutConverter(
            @NotNull Map<StructField, SearchField> fieldMap
    ) {

        String description = fieldMap.entrySet().stream().map(
                entry -> String.format(
                        "%s [spark type %s, search type %s]",
                        entry.getKey().name(),
                        entry.getKey().dataType(),
                        entry.getValue().getType()
                        )
                ).collect(Collectors.joining(", "));

        String message = String.format(
                "Could not find a suitable converter for %s fields (%s)",
                fieldMap.size(),
                description
        );

        return new SchemaCompatibilityException(message);
    }
}
